### PR TITLE
[IB method] Revised ghost point method

### PR DESCRIPTION
**Pull request by ktoepperwien**
IB surfaces may now be provided as point clouds in a given dimension `dim`. Surface normals are approximated through first order differences. Neumann boundary conditions now preserve 2nd-order accuracy, but only valid for isothermal flows. Density update at the ghost point not yet considered (see [Luo et al., Int. J. Heat Mass Transfer, 2016](https://linkinghub.elsevier.com/retrieve/pii/S0017931015009679)).
All required tensors for ghost point method now automatically initialized by calling `generate_initial_states` in `immersed_boundary_method.py`.

- [ ] `generate_custom_terrain_profile` in `terrain_utils.py` needs to be refactored (avoid `profile` string), or even removed
- [ ] Appropriate density treatment for non-isothermal (reacting) flows

### CODE REVIEW

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 1]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
 
 
+def init_ghost_point_helper_vars(
+    params: parameters_lib.SwirlLMParameters,
+    surface: tf.Tensor,
+    ib_interior_mask: FlowFieldVal,
+    dim: int,
+    my_var: str,
+    n_cores: initializer.ThreeIntTuple,
+) -> InitFn:
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:338]: Let's return the result dictionary, e.g. {'a': tf.Tensor(...), 'b': tf.Tensor(...), ...} directly from this function instead of an `init_fn` for each of these helper variables. We can reuse variables we computed earlier inside this function to prevent repeated operations. We'll need to call `partial_mesh_for_core` for each variable that depends on the coordinates inside this function.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 2]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+    z_ip2 = tf.clip_by_value(z_ip2, 0.0, params.lz)
+
+    # Get p,q,s indices of the lower left node of the cell enclosing the image
+    # point
+    def get_local_index(
+        ip_coord: tf.Tensor,
+        h: float,
+        n: int,
+        dim: int,
+        c: int,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:564]: Let's remove this parameter as it's no longer used.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 3]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+
+      Returns:
+        The local indices of the image points along a given coordinate.
+      """
+      index = tf.nest.map_structure(
+          lambda xi: tf.cast(xi // h, dtype=tf.int32) % (n - 2 * halo_width),
+          ip_coord,
+      )
+
+      ## index = tf.cond(
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:586]: Also remove the unused code.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 4]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+      """Computes the interpolation weights of each grid node enclosing a given
+      image point."""
+
+      node_dist = []
+      temp_mask = []
+
+      for i, j, k in itertools.product(range(2), range(2), range(2)):
+        p_x = common_ops.gather(xx,
+            tf.stack([
+                lower_left_index[0] + i,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:636]: Let's compute these 3 indices once only.

corner_index = tf.stack([lower_left_index[0] + i, lower_left_index[1] + j,  lower_left_index[2] + k], axis=1)

Reuse corner index everywhere below.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 5]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+      # exactly 8 (since there are 8 enclosing nodes in 3D space for an image
+      # point). However, if one or more grid nodes are inside the solid domain,
+      # the corresponding sum of `fluid_corner_point_mask` will be less than 8,
+      # which indicates that we have a degenerated cell. Note that
+      # `fluid_corner_point_mask` already flags every node with `1` if inside the
+      # fluid domain, and `0` if inside the solid domain.
+      # Source: Iaccarino and Verzicco, Applied Mechanics Reviews 56(3), 2003.
+      degenerated_cell_mask = tf.nest.map_structure(
+          lambda v: tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v)),
+          tf.math.reduce_max(fluid_corner_point_mask, axis=0)
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:696]: This should be reduce_sum❗

v = tf.math.reduce_sum(tf.stack(fluid_corner_point_mask))
degenerated_cell_mask = tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v))

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 6]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+      # the IB to `node_dist` which will now have 9 rows (8 distances between the
+      # image point and its surrounding grid nodes, and 1 additional row for the
+      # distance between the image point and the IB). However, the 9th row will
+      # only have non-zero distances, if the enclosing grid cell is degenerated,
+      # otherwise, the distance will be zero. As we will always use all 9 rows of
+      # `node_dist` to compute the interpolated value on the image point, we want
+      # to make sure that the BI is excluded for an eclosing cell that has all 8
+      # grid nodes in the fluid domain, and included otherwise.
+      node_dist.append(
+          tf.nest.map_structure(
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:712]: We don't need the map_structure here as both gp_dist and degenerated_cell_mask are 1D tf.Tensor.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 7]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+            lambda d, m: tf.math.pow(
+                tf.math.divide(1.0, d + _EPSILON), _EXP) * m,
+            gp_dist,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:715]: Let's make a note here that gp_dist is used because it's the same as the ip_dist.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 8]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+        node_dist.append(
+            tf.math.sqrt(
+                (image_point[0] - p_x) ** 2 +
+                (image_point[1] - p_y) ** 2 +
+                (image_point[2] - p_z) ** 2))
+
+        # Flag corner points with 1 which are in the fluid domain, and
+        # set all corner points to zero in the solid domain
+        temp_mask.append(
+            common_ops.gather(tf.transpose(ib_interior_mask, perm=[1, 2, 0]),
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:665]: `ib_interior_mask` contains halos if it's generated from a previous call to partial_mesh_for_core. So I believe what we should use here is ib_interior_mask[halo_width:-halo_width, halo_width:-halo_width, halo_width:-halo_width].

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 9]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+      # image point. Consequently, we need a separate set of interpolation weights
+      # for Neumann boundary conditions.
+      # Reference: R. Franke, Scattered data interpolation: tests of some
+      # methods, Mathematics of Computation 38(157), 1982,
+      # https://www.ams.org/mcom/1982-38-157/S0025-5718-1982-0637296-4/
+      # And:
+      # Cai et al., JCP 429, 2021, specifically Eq. 35, p. 9
+
+      # For Dirichlet boundary conditions:
+      gamma = tf.nest.map_structure(
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:746]: Let's make the gamma computation a helper function at the file level so that we can test it with artificial inputs.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 10]
```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
...
+        f'w_neumann{n}': convert_to_3d_tensor(w_n[n], ijk_gp,
+        ) for n in range(8)
+    })
+    output.update({
+        f'w_neumann_2x{n}': convert_to_3d_tensor(w_n_2x[n], ijk_gp,
+        ) for n in range(8)
+    })
+    output.update(
+        {
+            'nx': unit_normals_masked[..., 0],
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:809]: Let's try to eliminate the unused ones (or only keep then with the presence of a debugging flag).
